File: metrics.py
"""metrics.py — latency / throughput / batch instrumentation for experiments.

Attaches two pad probes and writes one CSV row per processed batch:

  * nvstreammux **src** pad  -> stamps a monotonic time when a batch is pushed.
  * nvtracker  **src** pad  -> the batch has been inferred + tracked; here we
      - pop the mux stamp (FIFO; the pipeline is linear so order is preserved)
        to get the compute latency (mux -> tracker),
      - use the buffer PTS vs the pipeline running clock for an e2e latency
        (capture -> here),
      - count how many frames were actually in the batch (== the timeout / skip
        effect) and how many detections each camera produced,
      - track newly-appearing track IDs as a cheap tracking-stability proxy.

Aggregation (mean / p50 / p99 / FES) is done offline by scripts/analyze.py so the
hot path stays cheap. What we log is raw per-batch rows.

No ground truth is available for live webcams, so "accuracy" here is a *proxy*
(track-ID stability). Swap in real mAP if you replay a labelled dataset.
"""

from __future__ import annotations

import csv
import os
import time
from typing import Optional, Set

import gi

gi.require_version("Gst", "1.0")
from gi.repository import Gst  # noqa: E402

import pyds  # noqa: E402

from detection_parser import parse_batch_meta


class MetricsCollector:
    """Collects per-batch metrics and streams them to a CSV file."""

    def __init__(self, csv_path: str, num_cams: int) -> None:
        self.num_cams = num_cams
        self._csv_path = csv_path
        self._pipeline: Optional[Gst.Pipeline] = None

        # State updated by the controllers each tick.
        self._active: Set[int] = set(range(num_cams))
        self._timeout_us: int = 0
        self._mux_batch: int = num_cams

        # {batch PTS -> mux-src monotonic time}. Keyed by PTS (not a FIFO) so a
        # dropped buffer between mux and tracker can't permanently desync the
        # compute-latency pairing (the FIFO version was one-off-forever fragile).
        self._mux_pts = {}
        # Per-camera map {frame PTS -> source arrival time}. Correlating by PTS
        # (not FIFO) survives nvstreammux dropping frames in live mode. Lets us
        # measure the TRUE latency a frame experiences (batch-wait + compute),
        # which the batch's own PTS misses because nvstreammux re-timestamps it.
        self._src_pts = {i: {} for i in range(num_cams)}
        self._src_pts_cap = 600
        self._batch_idx = 0

        # Tracking-stability proxy: distinct (camera, track_id) ever seen.
        self._seen_ids: Set = set()
        self._new_ids_cum = 0

        self._file = open(csv_path, "w", newline="")
        self._writer = csv.writer(self._file)
        self._writer.writerow(
            ["batch_idx", "t_mono", "n_in_batch", "n_real", "n_active", "active_mask",
             "timeout_us", "mux_batch", "compute_ms", "e2e_ms", "total_dets", "new_ids_cum"]
            + [f"dets_cam{i}" for i in range(num_cams)]
        )
        self._t_start = time.monotonic()

        # Opt-in diagnostic (default OFF; no effect unless DIAG_SOURCES is set):
        # per batch, dump which camera source_ids landed together and their PTS
        # skew (ms). Used to explain why sync-inputs caps n_in_batch below num_cams.
        self._diag = None
        diag_path = os.environ.get("DIAG_SOURCES")
        if diag_path:
            self._diag = open(diag_path, "w", newline="")
            self._diag.write("batch_idx,t_mono,n,source_ids,pts_skew_ms,per_cam_offset_ms\n")

    # ---- called by the controllers each control tick --------------------- #
    def set_active_cameras(self, active: Set[int]) -> None:
        self._active = set(active)

    def set_timeout_us(self, us: int) -> None:
        self._timeout_us = int(us)

    def set_mux_batch(self, bs: int) -> None:
        self._mux_batch = int(bs)

    # ---- probe wiring ---------------------------------------------------- #
    def attach(self, pipeline: Gst.Pipeline) -> None:
        """Attach the mux-src and tracker-src probes."""
        self._pipeline = pipeline
        mux = pipeline.get_by_name("stream-muxer")
        tracker = pipeline.get_by_name("tracker")
        if mux is None or tracker is None:
            raise RuntimeError("MetricsCollector: could not find stream-muxer / tracker.")
        mux.get_static_pad("src").add_probe(Gst.PadProbeType.BUFFER, self._mux_probe, None)
        tracker.get_static_pad("src").add_probe(Gst.PadProbeType.BUFFER, self._tracker_probe, None)

        # Per-camera arrival stamps at each source bin's (post-valve) src pad.
        for i in range(self.num_cams):
            src_bin = pipeline.get_by_name(f"source-bin-{i}")
            pad = src_bin.get_static_pad("src") if src_bin is not None else None
            if pad is not None:
                pad.add_probe(Gst.PadProbeType.BUFFER, self._make_src_probe(i), None)

    def _make_src_probe(self, cam_id: int):
        def _probe(pad, info, _u) -> Gst.PadProbeReturn:
            buf = info.get_buffer()
            if buf is not None and buf.pts != Gst.CLOCK_TIME_NONE:
                d = self._src_pts[cam_id]
                d[buf.pts] = time.monotonic()
                if len(d) > self._src_pts_cap:  # bound memory: drop oldest
                    del d[next(iter(d))]
            return Gst.PadProbeReturn.OK
        return _probe

    def _mux_probe(self, pad, info, _u) -> Gst.PadProbeReturn:
        buf = info.get_buffer()
        if buf is not None and buf.pts != Gst.CLOCK_TIME_NONE:
            self._mux_pts[buf.pts] = time.monotonic()
            if len(self._mux_pts) > self._src_pts_cap:  # bound memory: drop oldest
                del self._mux_pts[next(iter(self._mux_pts))]
        return Gst.PadProbeReturn.OK

    def _tracker_probe(self, pad, info, _u) -> Gst.PadProbeReturn:
        buf = info.get_buffer()
        if buf is None:
            return Gst.PadProbeReturn.OK
        now = time.monotonic()

        # compute latency (mux src -> tracker src): inference + tracking only.
        # Matched by batch PTS -> robust to any dropped buffer between the two pads.
        t_mux = self._mux_pts.pop(buf.pts, None) if buf.pts != Gst.CLOCK_TIME_NONE else None
        compute_ms = (now - t_mux) * 1e3 if t_mux is not None else -1.0

        # detections per camera + true source->output latency + tracking proxy.
        batch_meta = pyds.gst_buffer_get_nvds_batch_meta(hash(buf))
        per_cam = [0] * self.num_cams
        total = 0
        max_e2e = -1.0
        n_real = 0  # frames that genuinely arrived from a live camera this cycle
        diag_srcs = []  # (source_id, buf_pts) present this batch — for DIAG_SOURCES
        if batch_meta is not None:
            for frame in parse_batch_meta(batch_meta):
                cid = frame.camera_id
                if self._diag is not None:
                    diag_srcs.append((cid, frame.buf_pts))
                total += len(frame.detections)
                if 0 <= cid < self.num_cams:
                    per_cam[cid] = len(frame.detections)
                    # A frame is REAL iff its PTS matches a source arrival stamp.
                    # The legacy mux pads a skipped batch with REPEATED frames whose
                    # PTS was already consumed -> they won't match -> excluded. This
                    # gives an honest frame count (num_frames_in_batch overcounts the
                    # phantom repeats), and the frame's true source->output latency.
                    arrival = self._src_pts[cid].pop(frame.buf_pts, None)
                    if arrival is not None:
                        n_real += 1
                        lat_ms = (now - arrival) * 1e3
                        if lat_ms > max_e2e:
                            max_e2e = lat_ms
                for det in frame.detections:
                    if det.track_id >= 0:
                        key = (cid, det.track_id)
                        if key not in self._seen_ids:
                            self._seen_ids.add(key)
                            self._new_ids_cum += 1
        # a batch is only as fresh as its OLDEST frame -> worst-case in-batch e2e
        e2e_ms = max_e2e

        # n_in_batch = what the mux REPORTS (may include phantom repeats when
        # skipping); n_real = frames that actually arrived from a camera.
        n_in_batch = batch_meta.num_frames_in_batch if batch_meta is not None else 0

        # per-camera active bitmask, e.g. "1100" = cameras 0,1 active (2,3 skipped)
        active_mask = "".join("1" if i in self._active else "0" for i in range(self.num_cams))
        self._writer.writerow(
            [self._batch_idx, f"{now - self._t_start:.4f}", n_in_batch, n_real,
             len(self._active), active_mask, self._timeout_us, self._mux_batch,
             f"{compute_ms:.3f}", f"{e2e_ms:.3f}", total, self._new_ids_cum]
            + per_cam
        )
        if self._diag is not None and diag_srcs:
            ids = sorted(s for s, _ in diag_srcs)
            pts = [p for _, p in diag_srcs if p]
            skew_ms = (max(pts) - min(pts)) / 1e6 if len(pts) > 1 else 0.0
            # per-camera PTS offset from the batch's min PTS (ms), so we can SEE the
            # actual inter-camera timestamp structure: e.g. "0=+0.0|1=+8.3|2=+16.9".
            base = min(pts) if pts else 0
            per = "|".join(f"{s}=+{(p - base)/1e6:.1f}" for s, p in sorted(diag_srcs))
            self._diag.write(f"{self._batch_idx},{now - self._t_start:.4f},"
                             f"{len(ids)},{'|'.join(map(str, ids))},{skew_ms:.2f},{per}\n")
        self._batch_idx += 1
        return Gst.PadProbeReturn.OK

    def close(self) -> None:
        """Flush the CSV and print a one-line summary."""
        try:
            self._file.flush()
            self._file.close()
            if self._diag is not None:
                self._diag.flush()
                self._diag.close()
        except Exception:
            pass
        dur = max(1e-6, time.monotonic() - self._t_start)
        print(
            f"[metrics] wrote {self._batch_idx} batches to {self._csv_path} "
            f"({self._batch_idx / dur:.1f} batches/s over {dur:.1f}s; "
            f"{self._new_ids_cum} distinct tracks).",
            flush=True,
        )

File: cpp/src/metrics.cpp
#include "metrics.hpp"

#include <glib.h>

#include <algorithm>
#include <stdexcept>

#include "detection_parser.hpp"
#include "gstnvdsmeta.h"

namespace mcrt {

namespace {
double mono_secs() { return g_get_monotonic_time() / 1e6; }

/* Evict the oldest entry (smallest PTS — timestamps rise monotonically). */
void cap_map(std::map<guint64, double>& m, std::size_t cap) {
  while (m.size() > cap) m.erase(m.begin());
}
}  // namespace

MetricsCollector::MetricsCollector(const std::string& csv_path, int num_cams,
                                   int timeout_us, int mux_batch, bool sync_on)
    : num_cams_(num_cams),
      timeout_us_(timeout_us),
      mux_batch_(mux_batch),
      sync_on_(sync_on),
      csv_path_(csv_path),
      src_pts_(num_cams) {
  file_ = std::fopen(csv_path.c_str(), "w");
  if (file_ == nullptr)
    throw std::runtime_error("MetricsCollector: cannot open CSV for writing: " +
                             csv_path);
  std::fputs(
      "batch_idx,t_mono,n_in_batch,n_real,n_active,active_mask,timeout_us,"
      "mux_batch,compute_ms,e2e_ms,total_dets,new_ids_cum",
      file_);
  for (int i = 0; i < num_cams_; ++i) std::fprintf(file_, ",dets_cam%d", i);
  std::fputs(",drops_cum,arrivals_cum\n", file_);
  t_start_ = mono_secs();
}

MetricsCollector::~MetricsCollector() { close(); }

void MetricsCollector::attach(GstElement* pipeline) {
  GstElement* mux = gst_bin_get_by_name(GST_BIN(pipeline), "stream-muxer");
  GstElement* tracker = gst_bin_get_by_name(GST_BIN(pipeline), "tracker");
  if (mux == nullptr || tracker == nullptr) {
    if (mux) gst_object_unref(mux);
    if (tracker) gst_object_unref(tracker);
    throw std::runtime_error(
        "MetricsCollector: could not find stream-muxer / tracker.");
  }

  GstPad* mux_src = gst_element_get_static_pad(mux, "src");
  gst_pad_add_probe(mux_src, GST_PAD_PROBE_TYPE_BUFFER, mux_probe, this, nullptr);
  gst_object_unref(mux_src);

  GstPad* trk_src = gst_element_get_static_pad(tracker, "src");
  gst_pad_add_probe(trk_src, GST_PAD_PROBE_TYPE_BUFFER, tracker_probe, this,
                    nullptr);
  gst_object_unref(trk_src);

  /* The NEW mux emits "dropped" when sync-inputs discards a late frame —
   * count them (the legacy mux has no such signal; guard the lookup). */
  if (g_signal_lookup("dropped", G_OBJECT_TYPE(mux)) != 0)
    g_signal_connect(mux, "dropped", G_CALLBACK(on_dropped), this);

  for (int i = 0; i < num_cams_; ++i) {
    const std::string name = "source-bin-" + std::to_string(i);
    GstElement* bin = gst_bin_get_by_name(GST_BIN(pipeline), name.c_str());
    if (bin == nullptr) continue;
    GstPad* pad = gst_element_get_static_pad(bin, "src");
    if (pad != nullptr) {
      auto ctx = std::make_unique<SrcCtx>();
      ctx->self = this;
      ctx->cam_id = i;
      gst_pad_add_probe(pad, GST_PAD_PROBE_TYPE_BUFFER, src_probe, ctx.get(),
                        nullptr);
      src_ctxs_.push_back(std::move(ctx));
      gst_object_unref(pad);
    }
    gst_object_unref(bin);
  }

  gst_object_unref(mux);
  gst_object_unref(tracker);
}

GstPadProbeReturn MetricsCollector::src_probe(GstPad*, GstPadProbeInfo* info,
                                              gpointer user_data) {
  auto* ctx = static_cast<SrcCtx*>(user_data);
  GstBuffer* buf = GST_PAD_PROBE_INFO_BUFFER(info);
  if (buf != nullptr && GST_BUFFER_PTS_IS_VALID(buf)) {
    MetricsCollector* self = ctx->self;
    self->arrivals_cum_.fetch_add(1);
    std::lock_guard<std::mutex> lock(self->mu_);
    auto& m = self->src_pts_[ctx->cam_id];
    m[GST_BUFFER_PTS(buf)] = mono_secs();
    cap_map(m, kPtsCap);
  }
  return GST_PAD_PROBE_OK;
}

GstPadProbeReturn MetricsCollector::mux_probe(GstPad*, GstPadProbeInfo* info,
                                              gpointer user_data) {
  auto* self = static_cast<MetricsCollector*>(user_data);
  GstBuffer* buf = GST_PAD_PROBE_INFO_BUFFER(info);
  if (buf != nullptr && GST_BUFFER_PTS_IS_VALID(buf)) {
    std::lock_guard<std::mutex> lock(self->mu_);
    self->mux_pts_[GST_BUFFER_PTS(buf)] = mono_secs();
    cap_map(self->mux_pts_, kPtsCap);
  }
  return GST_PAD_PROBE_OK;
}

GstPadProbeReturn MetricsCollector::tracker_probe(GstPad*,
                                                  GstPadProbeInfo* info,
                                                  gpointer user_data) {
  auto* self = static_cast<MetricsCollector*>(user_data);
  GstBuffer* buf = GST_PAD_PROBE_INFO_BUFFER(info);
  if (buf != nullptr) self->handle_tracker_buffer(buf);
  return GST_PAD_PROBE_OK;
}

void MetricsCollector::on_dropped(GstElement*, gpointer, gpointer user_data) {
  static_cast<MetricsCollector*>(user_data)->drops_cum_.fetch_add(1);
}

/* The heart of the collector: runs once per batch on the tracker-src
 * streaming thread and writes one CSV row. Latencies are computed by looking
 * up the stamps the earlier probes stored, KEYED BY PTS — a frame the mux
 * dropped simply never gets its stamp consumed (and is eventually evicted by
 * cap_map), so drops can't desynchronize the pairing the way a FIFO would. */
void MetricsCollector::handle_tracker_buffer(GstBuffer* buf) {
  const double now = mono_secs();
  NvDsBatchMeta* batch_meta = gst_buffer_get_nvds_batch_meta(buf);
  const auto frames = parse_batch_meta(batch_meta);

  std::lock_guard<std::mutex> lock(mu_);
  if (file_ == nullptr) return;  // closed while a buffer was in flight

  /* compute latency (mux src -> tracker src), matched by the batch PTS. */
  double compute_ms = -1.0;
  if (GST_BUFFER_PTS_IS_VALID(buf)) {
    auto it = mux_pts_.find(GST_BUFFER_PTS(buf));
    if (it != mux_pts_.end()) {
      compute_ms = (now - it->second) * 1e3;
      mux_pts_.erase(it);
    }
  }

  /* Per-camera detections + true source->output latency + stability proxy. */
  std::vector<int> per_cam(num_cams_, 0);
  int total = 0;
  double max_e2e = -1.0;
  int n_real = 0;  // frames whose PTS matches a real source arrival stamp
  for (const auto& frame : frames) {
    const int cid = frame.camera_id;
    total += static_cast<int>(frame.detections.size());
    if (cid >= 0 && cid < num_cams_) {
      per_cam[cid] = static_cast<int>(frame.detections.size());
      auto& m = src_pts_[cid];
      auto it = m.find(frame.buf_pts);
      if (it != m.end()) {
        ++n_real;
        const double lat_ms = (now - it->second) * 1e3;
        if (lat_ms > max_e2e) max_e2e = lat_ms;
        m.erase(it);
      }
    }
    /* Tracking-stability proxy: count each (camera, track_id) pair the first
     * time it is seen. A pipeline variant that fragments tracks (e.g. by
     * dropping frames) shows up as a faster-growing new_ids_cum. */
    for (const auto& det : frame.detections) {
      if (det.track_id >= 0 &&
          seen_ids_.insert({cid, det.track_id}).second)
        ++new_ids_cum_;
    }
  }

  const int n_in_batch =
      batch_meta != nullptr ? static_cast<int>(batch_meta->num_frames_in_batch) : 0;

  /* No skipping in this app: all cameras are always active. */
  const std::string active_mask(num_cams_, '1');
  std::fprintf(file_, "%ld,%.4f,%d,%d,%d,%s,%d,%d,%.3f,%.3f,%d,%ld",
               batch_idx_, now - t_start_, n_in_batch, n_real, num_cams_,
               active_mask.c_str(), timeout_us_, mux_batch_, compute_ms,
               max_e2e, total, new_ids_cum_);
  for (int i = 0; i < num_cams_; ++i) std::fprintf(file_, ",%d", per_cam[i]);
  std::fprintf(file_, ",%ld,%ld\n", drops_cum_.load(), arrivals_cum_.load());
  ++batch_idx_;
  real_frames_cum_ += n_real;
}

void MetricsCollector::close() {
  std::lock_guard<std::mutex> lock(mu_);
  if (closed_) return;
  closed_ = true;
  if (file_ != nullptr) {
    std::fflush(file_);
    std::fclose(file_);
    file_ = nullptr;
  }
  const double dur = std::max(1e-6, mono_secs() - t_start_);
  const long arrived = arrivals_cum_.load();
  std::fprintf(stderr,
               "[metrics] wrote %ld batches to %s (%.1f batches/s over %.1fs; "
               "%ld distinct tracks; sync %s; processed %ld of %ld arrived "
               "frames; mux dropped-signal count %ld).\n",
               batch_idx_, csv_path_.c_str(), batch_idx_ / dur, dur,
               new_ids_cum_, sync_on_ ? "ON" : "off", real_frames_cum_,
               arrived, drops_cum_.load());
}

}  // namespace mcrt
